apps/backend/app/core/algorithms.py

```
@@ -0,0 +1,479 @@
+        tfidf_matrix = vectorizer.transform([documents[doc_idx]])
+        
+        # Get top keywords
+        scores = []
+        for token, idx in vectorizer.vocabulary_.items():
+            score = tfidf_matrix[0, idx]
+            if score > 0:
+                scores.append((token, score))
+        
+        return heapq.nlargest(top_k, scores, key=lambda x: x[1])
+    
+    @lru_cache(maxsize=1000)
```

Ruff (0.11.9)

Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks

(B019)